source/SVInt.cpp

```diff
@@ -476,6 +476,85 @@ SVInt SVInt::lshr(uint32_t amount) const {
     return result;
 }
 
+SVInt SVInt::ashr(const SVInt& rhs) const {
+    if (!signFlag)
+        return lshr(rhs);
+    if (rhs.hasUnknown())
+        return createFillX(bitWidth, signFlag);
+    if (rhs >= bitWidth)
+        return SVInt(bitWidth, *this >= 0 ? 0 : -1, signFlag);
+
+    uint16_t contractedWidth = bitWidth - rhs.getAssertUInt64();
+    SVInt tmp = lshr(rhs);
+
+    if (contractedWidth <= 64 && getNumWords() > 1) {
+        // signExtend won't be safe as it will assume it is operating on
+        // a single-word input when it isn't
+        // so let's manually take care of that case here.
+
+        uint64_t* newData = new uint64_t[getNumWords()]();
+        // get sign-extended value
+        uint64_t newVal = tmp.pVal[0] << (SVInt::BITS_PER_WORD - contractedWidth);
+        newData[0] = (int64_t)newVal >> (SVInt::BITS_PER_WORD - contractedWidth);
+        for (size_t i = 1; i < getNumWords(); ++i) {
+            // sign extend the rest based on original sign
+            newData[i] = val & (1L << 63) ? 0 : ~0L;
+        }
+        SVInt ret(newData, bitWidth, signFlag, false);
+        ret.clearUnusedBits();
+        return ret;
+    }
+    // Pretend our width is just the width we shifted to, then signExtend
+    tmp.bitWidth = contractedWidth;
+    return signExtend(tmp, bitWidth);
+}
+
+SVInt SVInt::ashr(uint32_t amount) const {
+    if (!signFlag)
+        return lshr(amount);
+    if (amount == 0)
+        return *this;
+    if (amount >= bitWidth)
+        return SVInt(bitWidth, *this >= 0 ? 0 : -1, signFlag);
+
+    uint16_t contractedWidth = bitWidth - amount;
+    SVInt tmp = lshr(amount);
+    tmp.bitWidth = contractedWidth;
+    return signExtend(tmp, bitWidth);
```

Comment:
oops, I fixed the SVInt version and in so doing totally forgot that the uint32_t version existed. I will make one call the other.